src/core/transaction.c

```diff
@@ -1047,27 +1047,42 @@ int transaction_add_job_and_dependencies(
                 }
 
                 if (IN_SET(type, JOB_STOP, JOB_RESTART)) {
-                        UnitDependencyAtom atom;
-                        JobType ptype;
-
                         /* We propagate STOP as STOP, but RESTART only as TRY_RESTART, in order not to start
                          * dependencies that are not around. */
-                        if (type == JOB_RESTART) {
-                                atom = UNIT_ATOM_PROPAGATE_RESTART;
-                                ptype = JOB_TRY_RESTART;
-                        } else {
-                                ptype = JOB_STOP;
-                                atom = UNIT_ATOM_PROPAGATE_STOP;
-                        }
 
-                        UNIT_FOREACH_DEPENDENCY(dep, ret->unit, atom) {
-                                JobType nt;
+                        propagated_restart = set_new(NULL);
+                        if (!propagated_restart)
+                                return -ENOMEM;
+
+                        if (type == JOB_RESTART)
+                                UNIT_FOREACH_DEPENDENCY(dep, ret->unit, UNIT_ATOM_PROPAGATE_RESTART) {
+                                        JobType nt;
+
+                                        r = set_put(propagated_restart, dep);
```

Comment:
please use `set_ensure_put()` here. and drop the the set_new() call above. THis will mean we'll allocate the set lazily once we need it. shortens the code a bit.

all our code treats NULL sets as empty sets, hence this is pretty effective